docs/configuration.md
# Configuration

<!-- content start -->
This are the configuration settings for the `/etc/unipi/control.yaml`.

## Device

| Key              | Value                                                                                                                                                               |            |
|------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------|
| `name`           | The device name for the subscribe and publish topics. Default is the hostname. This name must be unique. This is important if multiple Unipi devices are available. | optionally |
| `suggested_area` | Used as suggested area for the device in Home Assistant.                                                                                                            | optionally |


```yaml
# control.yaml
device_info:
  name: Unipi
  suggested_area: Distribution Cabinet
```

## MQTT

| Key                  | Value                                                                                                                                                                                                                  |
|----------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `host`               | The hostname or IP address of the remote broker: Default is `localhost`.                                                                                                                                               |
| `port`               | The network port of the server host to connect to. Defaults is `1883`.                                                                                                                                                 |
| `keepalive`          | Maximum period in seconds allowed between communications with the broker. If no other messages are being exchanged, this controls the rate at which the client will send ping messages to the broker. Default is `15`. |
| `retry_limit`        | Number of attempts to connect to the MQTT broker. Default to `30` (Disable with `False`).                                                                                                                              |
| `reconnect_interval` | Time between connection attempts. Default is `10`.                                                                                                                                                                     |
| `username`           | Username to use for MQTT authentication. Default is `None` (for brokers without authentication set).                                                                                                                   |
| `password`           | Password to use for MQTT authentication. Default is `None` (for brokers without authentication set).                                                                                                                   |

```yaml
# control.yaml
mqtt:
  host: localhost
  port: 1883
  connection:
    keepalive: 15
    retry_limit: 30
    reconnect_interval: 10
```

## Modbus

### TCP

| Key             | Value                                            |
|-----------------|--------------------------------------------------|
| `host`          | The modbus TCP host. Default is `localhost`.     |
| `port`          | The modbus TCP port. Default is `502`.           |
| `scan_interval` | Modbus register scan interval. Default is `0.2`. |


```yaml
# control.yaml
modbus_tcp:
  host: localhost
  port: 502
```

### Serial

| Key                       | Value                                                 |
|---------------------------|-------------------------------------------------------|
| `port`                    | The modbus RTU device. Default is `/dev/extcomm/0/0`. |
| `baud_rate`               | The baud rate for modbus RTU. Default is `2400`.      |
| `parity`                  | The parity for modbus RTU. Default is `N`.            |
| `scan_interval`           | Modbus register scan interval. Default is `0.2`.      |
| `unit`                    | A list of all modbus RTU devices.                     |
| `unit` » `unit`           | The unique modbus RTU unit ID.                        |
| `unit` » `device_name`    | Custom device name. Used for the Home Assistant UI.   |
| `unit` » `suggested_area` | Used as entity area in Home Assistant.                |

```yaml
# control.yaml
modbus_serial:
  port: /dev/extcomm/0/0
  baud_rate: 9600
  parity: N
  units:
    - unit: 1
      device_name: Eastron SDM120M
      identifier: Eastron_SDM120M
      suggested_area: Workspace
```

## Home Assistant

| Key                | Value                                                           |
|--------------------|-----------------------------------------------------------------|
| `enabled`          | Enable Home Assistant MQTT Discovery. Default is `true`.        |
| `discovery_prefix` | The prefix for the discovery topic. Default is `homeassistant`. |

```yaml
# control.yaml
homeassistant:
  enabled: true
  discovery_prefix: homeassistant
```

## Features

It's possible to give the features friendly names. This names will be used for switches and binary sensors and sensors in Home Assistant.

| Key                    | Value                                                                                                               |            |
|------------------------|---------------------------------------------------------------------------------------------------------------------|------------|
| `object_id`            | Used as entity ID in Home Assistant.                                                                                | optionally |
| `friendly_name`        | Used as entity name in Home Assistant.                                                                              | optionally |
| `icon`                 | Used as icon in Home Assistant. Any icon from [materialdesignicons.com](https://materialdesignicons.com). Prefix name with mdi:, ie `mdi:home`. | optionally |
| `device_class`         | Used for [Device Class](https://www.home-assistant.io/docs/configuration/customizing-devices/#device-class) in Home Assistant.                                                                     | optionally |
| `state_class`          | Used for [State Class](https://developers.home-assistant.io/docs/core/entity/sensor/#available-state-classes) in Home Assistant. Thid is only for sensors.                                            | optionally |
| `unit_of_measurement`  | Used as measurement unit in Home Assistant. Only for sensors.                                                       | optionally |
| `invert_state`         | Invert the `ON`/`OFF` state. Default is `false`. Only for binary sensors.                                           | optionally |

```yaml
# control.yaml
features:
  di_3_02:
    object_id: workspace_switch_up
    friendly_name: Workspace - Switch up
  di_3_03:
    object_id: workspace_switch_down
    friendly_name: Workspace - Switch down
  active_power_1:
    object_id: workspace_active_power
    friendly_name: Workspace - Active Power
    device_class: power
    state_class: measurement
    unit_of_measurement: W
```

## Covers

> **Warning:** Driving both signals up and down at the same time can damage the motor. There are a couple of checks to prevent such situation. Use this software at your own risk! I do not take any responsibility for it!

The Home Assistant Discovery for the covers is optionally. Covers can be control with MQTT topics without Home Assistant.

| Key                | Value                                                                                                                                      |             |
|--------------------|--------------------------------------------------------------------------------------------------------------------------------------------|-------------|
| `object_id`        | Used as entity ID in Home Assistant.                                                                                                       |             |
| `friendly_name`    | Used as entity name in Home Assistant.                                                                                                     | optionally  |
| `device_class`     | Device class can be "awning", "curtain", "door", "garage", "gate", "shade", "blind", "shutter" or "window"                                 | optionally  |
| `cover_run_time`   | Define the time (in seconds) it takes for the cover to fully open or close.                                                                | optionally  |
| `tilt_change_time` | Define the time (in seconds) that the tilt changes from fully open to fully closed state. Tilt is only available for device class "blind". | optionally  |
| `cover_up`         | Output circuit name from a relay or digital output.                                                                                        |             |
| `cover_down`       | Output circuit name from a relay or digital output.                                                                                        |             |

```yaml
# control.yaml
covers:
  - id: workspace_1
    friendly_name: "Workspace - Blind 1"
    device_class: "blind"
    cover_run_time: 35.5
    tilt_change_time: 1.5
    cover_up: ro_3_03
    cover_down: ro_3_02
```

### Calibration

The covers save its cover and tilt position in a temporary file each time a stop command is executed. If the Unipi Control is interrupted, and a command (open/closing) is currently being executed, or the system is restarted, the calibration mode is enabled.
When the Unipi Control starts in calibration mode, the cover fully open and disable calibration mode. This is required for the correct cover and tilt position.

> Cover calibration mode is disabled when `persistent_tmp_dir` is enabled! To manualy reset the covers you must delete all files under `/var/tmp/unipi`.

## Advanced

| Key                  | Value                                                                                                                                                                                               |
|----------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `persistent_tmp_dir` | Enable persistent temporary directory. All temporary files are written to `/var/tmp/unipi` instead of `/tmp/unipi`. Default is `false`. If this is enabled then cover calibration mode is disabled! |

```yaml
# control.yaml
advanced:
  persistent_tmp_dir: True
```


## Logging

| Key                       | Value                                                                               |
|---------------------------|-------------------------------------------------------------------------------------|
| `level`                   | Set level to `debug`, `info`, `warning` or `error`. Default is `error`.             |
| `mqtt` » `meters_level`   | Show meter MQTT messages only for this log level or greater. Default is `error`.    |
| `mqtt` » `features_level` | Show features MQTT messages only for this log level or greater. Default is `error`. |
| `mqtt` » `covers_level`   | Show covers MQTT messages only for this log level or greater. Default is `error`.   |

```yaml
# control.yaml
logging:
  level: info
  mqtt:
    meters_level: error
    features_level: info
    covers_level: info
```

<!-- content end -->


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Added support for MQTT broker authentication with username and password
- Added support for Python 3.12
- Added support for Unipi Extensions
- Added a lot of debug log messages for better Modbus RTU/TCP debugging
- Added `scan_interval` to Modbus RTU/TCP configuration

### Changed

- Faster reading from Modbus RTU/TCP
- Use "Unipi" as name instead of "Neuron"
- Refactor deprecated aiomqtt code

### Changed

- Bump pymodbus to version 3.5.4

## [3.1.0] - 2023-10-03

### Added

- Added support for more Home Assistant cover [device classes](https://www.home-assistant.io/integrations/cover/).
- Added support for cover position to all [device classes](https://www.home-assistant.io/integrations/cover/).

### Changed

- Bump aiomqtt to version 1.2.1
- Bump pymodbus to version 3.5.2

## [3.0.0] - 2023-09-15

### Added

- Added `modbus_tcp` as a new section in the config file.
- Added MQTT logging options to set `meters_level`,  `features_level` and `covers_level` in the config file.

### Changed

- **BREAKING CHANGE** MQTT topics updated to work with Home Assistant 2023.8.0 ([#95159](https://github.com/home-assistant/core/pull/95159))
- **BREAKING CHANGE** in the config file: the section `modbus` was renamed to `modbus_serial`.
- Bump pymodbus to version 3.5.0
- Bump aiomqtt to version 1.2.0
- Bump build to version 1.0.0
- Bump pytest to version to 7.4.1
- Bump mypy to version 1.5.1
- Bump coverage to version 7.3.0
- Bump ruff to version 0.0.284 
- Updated [configuration.md](docs/configuration.md)

### Fixed

- Fixed mypy errors

## [2.1.0] - 2023-07-14

### Added

- Added `advanced` section in `control.yaml` with the option `persistent_tmp_dir`.
- Added documentation in [configuration.md](docs/configuration.md) for the `advanced` section.

### Changed

- Bump pymodbus to version 3.3.2
- Bump ayncio-mqtt to version 1.0.0 (the package name changed to aiomqtt)

### Fixed

- Fixed crashing Unipi Control service after reinstall/update `unipi-modbus-tools` OPKG package.
- Added `Restart=on-failure` for systemd service, in the OPKG package, to prevent unwanted crashes.
- Repair the OPKG package for the Unipi Control OS. Added the missing control files to the package. 

## [2.0.0] - 2023-06-23

### Added

- This CHANGELOG.md
- Added the developer guide to [CONTRIBUTING.md](CONTRIBUTING.md)
- Added `setuptools_scm` for automatic versioning.
- Added ruff as new linter.
- Added date and log level to unipi-control CLI output
- Added better error handling if hardware definition files are invalid

### Changed

- Changed version format to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
- Moved from legacy python package installer configuration file (setup.cfg) to [pyproject.toml](pyproject.toml)
- Updated pytest/coverage configuration
  - Save all output files e.g. `pytest.xml` in the reports folder 
- Split README.md content in separate files under [docs](docs) (Prepare for [docs.superbox.one](https://docs.superbox.one))
- Changed project structure (flat layout) to `unipi_control`, `data`, `scripts` and `tests` folder.

### Removed

- Removed superbox-utils dependencies.
- Removed old `flake8` linter.

### Fixed

- Fixed wrong logging level. Logging level from YAML configuration was not set correctly.

## [2023.7] - 2023-05-03

### Added

- New GitHub Action for deploy to Unipi Control OS.
  - Binary files to the release downloads.
  - Wheel file to the release downloads.
  - OPKG package to the release downloads (required for Unipi Control OS).
- New argument `--config` to `unipi-control` command for custom config path.
- `unipi-config-backup` command for easy creating configuration backups.
- `unipi-config-converter` to convert EVOK yaml files to Unipi Control yaml files.

### Changed

- Bump `superbox-utils` to `2023.6`.
- Updated list of supported devices.

## [2023.6] - 2023-03-19

### Added

- Added support for Modbus RTU.
- Added support for Eastron devices.

### Changed

- Update systemd service for Unipi Control OS.
- Updated packages version.
- Updated install process and [README.md](README.md).

### Fixed

- Fix missing pyserial package.


unipi_control/config.py
"""Read config from yaml file and create config data class."""

import dataclasses
import logging
import re
import socket
import typing
from dataclasses import dataclass
from dataclasses import field
from dataclasses import is_dataclass
from functools import cached_property
from pathlib import Path
from typing import Any
from typing import Dict
from typing import Final
from typing import Iterator
from typing import List
from typing import NamedTuple
from typing import Optional
from typing import Tuple
from typing import Type
from typing import Union

from aiomqtt.client import MQTT_LOGGER

from unipi_control.helpers.exceptions import ConfigError
from unipi_control.helpers.exceptions import YamlError
from unipi_control.helpers.log import LOG_FORMAT
from unipi_control.helpers.log import LOG_LEVEL
from unipi_control.helpers.log import LOG_NAME
from unipi_control.helpers.log import SIMPLE_LOG_FORMAT
from unipi_control.helpers.log import SystemdHandler
from unipi_control.helpers.yaml import yaml_loader_safe

UNIPI_LOGGER: logging.Logger = logging.getLogger(LOG_NAME)

DEFAULT_CONFIG_DIR: Final[Path] = Path("/etc/unipi")
DEVICE_CLASSES: Final[List[str]] = [
    "awning",
    "curtain",
    "door",
    "garage",
    "gate",
    "shade",
    "blind",
    "shutter",
    "window",
]

MODBUS_BAUD_RATES: Final[List[int]] = [2400, 4800, 9600, 19200, 38400, 57600, 115200]
MODBUS_PARITY: Final[List[str]] = ["E", "O", "N"]


class LogPrefix:
    CONFIG: Final[str] = "[CONFIG]"
    COVER: Final[str] = "[COVER]"
    DEVICEINFO: Final[str] = "[DEVICEINFO]"
    MQTT: Final[str] = "[MQTT]"
    FEATURE: Final[str] = "[FEATURE]"
    MODBUS: Final[str] = "[MODBUS]"


class RegexValidation(NamedTuple):
    regex: str
    error: str


class Validation:
    NAME: RegexValidation = RegexValidation(
        regex=r"^[A-Za-z\d\s_-]*$", error="The following characters are prohibited: a-z 0-9 -_ space"
    )

    ID: RegexValidation = RegexValidation(
        regex=r"^[A-Za-z\d_-]*$", error="The following characters are prohibited: a-z 0-9 -_"
    )


@dataclass
class ConfigLoaderMixin:
    @staticmethod
    def _update_field_with_dataclass(
        value: Union[str, int, List[Any], Dict[str, Any]], name: str, field_type: Type[Any]
    ) -> Union[str, int, List[Any], Dict[str, Any]]:
        field_origin: Any = typing.get_origin(field_type)
        field_args: Tuple[Any, ...] = typing.get_args(field_type)

        if field_origin == list and len(field_args) == 1:
            _list: List[Any] = []

            if isinstance(value, list):
                for list_item in value:
                    _dataclass = field_args[0]()
                    _dataclass.update(list_item)
                    _list.append(_dataclass)
            else:
                msg = f"Expected {name} to be {field_origin}, got {value!r}"
                raise ConfigError(msg)

            value = _list
        elif field_origin == dict:
            _dict: Dict[str, Any] = {}

            if isinstance(value, dict):
                for key, dict_value in value.items():
                    _dataclass = field_args[1]()
                    _dataclass.update(dict_value)
                    _dict[key] = _dataclass
            else:
                msg = f"Expected {name} to be {field_origin}, got {value!r}"
                raise ConfigError(msg)

            value = _dict

        return value

    def update(self, new: Dict[str, Any]) -> None:
        """Update and validate config data class with settings from a dictionary.

        Parameters
        ----------
        new: dict
            Overwrite settings as dictionary.
        """
        for key, value in new.items():
            if hasattr(self, key):
                item = getattr(self, key)

                if is_dataclass(item):
                    item.update(new=value)
                else:
                    _value = value

                    for _field in dataclasses.fields(self):
                        if _field.name == key:
                            _value = self._update_field_with_dataclass(value, name=_field.name, field_type=_field.type)
                            break

                    setattr(self, key, _value)

        self.validate()

    def update_from_yaml_file(self, config_file: Path) -> None:
        """Update and validate config data class with settings from a YAML file.

        Parameters
        ----------
        config_file: Path
            Path to the YAML file.
        """
        if config_file.is_file() and config_file.exists():
            try:
                yaml_data: Dict[str, Any] = yaml_loader_safe(config_file)
            except YamlError as error:
                raise ConfigError(error) from error

            if isinstance(yaml_data, dict):
                self.update(yaml_data)

    def validate(self) -> None:
        """Validate config data class arguments."""
        for _field in dataclasses.fields(self):
            value: Any = getattr(self, _field.name)
            field_type = typing.get_origin(_field.type) or _field.type

            if is_dataclass(value):
                value.validate()
            else:
                if method := getattr(self, f"_validate_{_field.name}", None):
                    setattr(self, _field.name, method(getattr(self, _field.name), name=_field.name))

                if not isinstance(value, field_type) and not is_dataclass(value):
                    msg = f"Expected {_field.name} to be {field_type}, got {value!r}"
                    raise ConfigError(msg)


@dataclass
class DeviceInfo(ConfigLoaderMixin):
    name: str = field(default=socket.gethostname())
    manufacturer: str = field(default="Unipi technology")
    suggested_area: str = field(default_factory=str)

    @staticmethod
    def _validate_name(value: str, name: str) -> str:
        if re.search(Validation.NAME.regex, value) is None:
            msg = f"{LogPrefix.DEVICEINFO} Invalid value '{value}' in '{name}'. {Validation.NAME.error}"
            raise ConfigError(msg)

        return value


@dataclass
class MqttConfig(ConfigLoaderMixin):
    host: str = field(default="localhost")
    port: int = field(default=1883)
    keepalive: int = field(default=15)
    retry_limit: int = field(default=30)
    reconnect_interval: int = field(default=10)
    username: str = field(default_factory=str)
    password: str = field(default_factory=str)


@dataclass
class FeatureConfig(ConfigLoaderMixin):
    object_id: str = field(default_factory=str)
    friendly_name: str = field(default_factory=str)
    icon: str = field(default_factory=str)
    device_class: str = field(default_factory=str)
    state_class: str = field(default_factory=str)
    unit_of_measurement: str = field(default_factory=str)
    suggested_area: str = field(default_factory=str)
    invert_state: bool = field(default=False)

    @staticmethod
    def _validate_object_id(value: str, name: str) -> str:
        value = value.lower()

        if re.search(Validation.ID.regex, value) is None:
            msg = f"{LogPrefix.FEATURE} Invalid value '{value}' in '{name}'. {Validation.ID.error}"
            raise ConfigError(msg)

        return value


@dataclass
class CoverConfig(ConfigLoaderMixin):
    object_id: str = field(default_factory=str)
    friendly_name: str = field(default_factory=str)
    suggested_area: str = field(default_factory=str)
    device_class: str = field(default_factory=str)
    cover_run_time: float = field(default_factory=float)
    tilt_change_time: float = field(default_factory=float)
    cover_up: str = field(default_factory=str)
    cover_down: str = field(default_factory=str)

    def __post_init__(self) -> None:
        self.cover_run_time = float(self.cover_run_time)

    def validate(self) -> None:
        """Validate cover configuration."""
        for _field in ("object_id", "friendly_name", "device_class", "cover_up", "cover_down"):
            if not getattr(self, _field):
                msg = f"{LogPrefix.COVER} Required key '{_field}' is missing! {self!r}"
                raise ConfigError(msg)

        super().validate()

    @staticmethod
    def _validate_object_id(value: str, name: str) -> str:
        value = value.lower()

        if re.search(Validation.ID.regex, value) is None:
            msg = f"{LogPrefix.COVER} Invalid value '{value}' in '{name}'. {Validation.ID.error}"
            raise ConfigError(msg)

        return value

    def _validate_device_class(self, value: str, name: str) -> str:
        if (value := value.lower()) not in DEVICE_CLASSES:
            exception_message: str = (
                f"{LogPrefix.COVER} Invalid value '{self.device_class}' in '{name}'. "
                f"The following values are allowed: {' '.join(DEVICE_CLASSES)}."
            )
            raise ConfigError(exception_message)

        return value


@dataclass
class ModbusUnitConfig(ConfigLoaderMixin):
    unit: int = field(default_factory=int)
    device_name: str = field(default_factory=str)
    identifier: str = field(default_factory=str)
    suggested_area: str = field(default_factory=str)

    def _validate_device_name(self, value: str, name: str) -> str:  # noqa: ARG002
        if not value:
            msg = f"{LogPrefix.MODBUS} Device name for unit '{self.unit}' is missing!"
            raise ConfigError(msg)

        return value


@dataclass
class ModbusTCPConfig(ConfigLoaderMixin):
    host: str = field(default="localhost")
    port: int = field(default=502)
    scan_interval: float = field(default=0.02)


@dataclass
class ModbusSerialConfig(ConfigLoaderMixin):
    port: str = field(default="/dev/extcomm/0/0")
    baud_rate: int = field(default=2400)
    parity: str = field(default="N")
    scan_interval: float = field(default=0.02)
    units: List[ModbusUnitConfig] = field(init=False, default_factory=list)

    def get_units_by_identifier(self, identifier: str) -> Iterator[ModbusUnitConfig]:
        """Filter units by identifier.

        Parameters
        ----------
        identifier: str
            A unique units identifier.

        Returns
        -------
        Generator:
            Filtered units.
        """
        return (modbus_unit for modbus_unit in self.units if modbus_unit.identifier == identifier)

    @staticmethod
    def _validate_units(value: List[ModbusUnitConfig], name: str) -> List[ModbusUnitConfig]:  # noqa: ARG004
        unique_units: List[int] = []

        for unit in value:
            if unit.unit in unique_units:
                msg = f"{LogPrefix.MODBUS} Duplicate modbus unit '{unit.unit}' found in 'units'!"
                raise ConfigError(msg)

            unique_units.append(unit.unit)

        return value

    @staticmethod
    def _validate_baud_rate(value: int, name: str) -> int:  # noqa: ARG004
        if value not in MODBUS_BAUD_RATES:
            exception_message: str = (
                f"{LogPrefix.MODBUS} Invalid baud rate '{value}'. "
                f"The following baud rates are allowed: {' '.join(str(baud_rate) for baud_rate in MODBUS_BAUD_RATES)}."
            )
            raise ConfigError(exception_message)

        return value

    @staticmethod
    def _validate_parity(value: str, name: str) -> str:
        if (value := value.upper()) not in MODBUS_PARITY:
            exception_message: str = (
                f"{LogPrefix.MODBUS} Invalid value '{value}' in '{name}'. "
                f"The following parity options are allowed: {' '.join(MODBUS_PARITY)}."
            )
            raise ConfigError(exception_message)

        return value


@dataclass
class HomeAssistantConfig(ConfigLoaderMixin):
    enabled: bool = field(default=True)
    discovery_prefix: str = field(default="homeassistant")

    def _validate_discovery_prefix(self, value: str, name: str) -> str:
        value = value.lower()

        if re.search(Validation.ID.regex, value) is None:
            exception_message: str = (
                f"[{self.__class__.__name__.replace('Config', '').upper()}] "
                f"Invalid value '{value}' in '{name}'. {Validation.ID.error}"
            )
            raise ConfigError(exception_message)

        return value


@dataclass
class AdvancedConfig(ConfigLoaderMixin):
    persistent_tmp_dir: bool = False


@dataclass
class MqttLoggingConfig(ConfigLoaderMixin):
    meters_level: str = field(default="error")
    features_level: str = field(default="error")
    covers_level: str = field(default="error")


@dataclass
class LoggingConfig(ConfigLoaderMixin):
    level: str = field(default="error")
    mqtt: MqttLoggingConfig = field(default_factory=MqttLoggingConfig)

    @property
    def verbose(self) -> int:
        """Get logging verbose level as integer."""
        return list(LOG_LEVEL).index(self.level)

    def init(
        self,
        log: Optional[str] = None,
        verbose: int = 0,
        fmt: Optional[str] = None,
    ) -> None:
        """Initialize logger handler and formatter.

        Parameters
        ----------
        log: str
            set log handler to systemd or stdout.
        verbose: int
            Logging verbose level as integer.
        fmt: str, optional
            Custom logging formatter
        """
        UNIPI_LOGGER.handlers.clear()
        UNIPI_LOGGER.setLevel(LOG_LEVEL[self.level])
        MQTT_LOGGER.handlers.clear()

        if log == "systemd":
            systemd_handler = SystemdHandler()
            systemd_handler.setFormatter(logging.Formatter(fmt or SIMPLE_LOG_FORMAT))
            UNIPI_LOGGER.addHandler(systemd_handler)
            MQTT_LOGGER.addHandler(systemd_handler)
        else:
            stdout_handler: logging.Handler = logging.StreamHandler()
            stdout_handler.setFormatter(logging.Formatter(fmt or LOG_FORMAT))
            UNIPI_LOGGER.addHandler(stdout_handler)
            MQTT_LOGGER.addHandler(stdout_handler)

        if verbose > 0:
            self.update_level(verbose)

    def update_level(self, verbose: int) -> None:
        """Update the logging level in config data class.

        Parameters
        ----------
        verbose: int
            Logging verbose level as integer.
        """
        levels: List[int] = list(LOG_LEVEL.values())
        level: int = levels[min(max(verbose, self.verbose), len(levels) - 1)]

        UNIPI_LOGGER.setLevel(level)

    def _validate_level(self, value: str, name: str) -> str:  # noqa: ARG002
        if (value := value.lower()) not in LOG_LEVEL:
            exception_message: str = (
                f"[{self.__class__.__name__.replace('Config', '').upper()}] "
                f"Invalid log level '{self.level}'. The following log levels are allowed: {' '.join(LOG_LEVEL.keys())}."
            )
            raise ConfigError(exception_message)

        return value


@dataclass
class Config(ConfigLoaderMixin):
    device_info: DeviceInfo = field(default_factory=DeviceInfo)
    mqtt: MqttConfig = field(default_factory=MqttConfig)
    modbus_tcp: ModbusTCPConfig = field(default_factory=ModbusTCPConfig)
    modbus_serial: ModbusSerialConfig = field(default_factory=ModbusSerialConfig)
    homeassistant: HomeAssistantConfig = field(default_factory=HomeAssistantConfig)
    features: Dict[str, FeatureConfig] = field(init=False, default_factory=dict)
    covers: List[CoverConfig] = field(init=False, default_factory=list)
    advanced: AdvancedConfig = field(default_factory=AdvancedConfig)
    logging: LoggingConfig = field(default_factory=LoggingConfig)
    config_base_dir: Path = field(default=DEFAULT_CONFIG_DIR)
    unipi_tmp_dir: Path = field(default=Path("/tmp/unipi"))
    sys_bus_dir: Path = field(default=Path("/sys/bus/i2c/devices"))

    @cached_property
    def hardware_dir(self) -> Path:
        """Return hardware path to Unipi devices and extensions."""
        return self.config_base_dir / "hardware"

    def __post_init__(self) -> None:
        self.unipi_tmp_dir.mkdir(exist_ok=True)
        self.update_from_yaml_file(config_file=self.config_base_dir / "control.yaml")

        if self.advanced.persistent_tmp_dir:
            self.unipi_tmp_dir = Path("/var/tmp/unipi")

        self._validate_feature_object_ids()
        self._validate_covers_circuits()
        self._validate_cover_ids()

    def _validate_feature_object_ids(self) -> None:
        object_ids: List[str] = []

        for feature in self.features.values():
            if feature.object_id in object_ids:
                msg = f"{LogPrefix.FEATURE} Duplicate ID '{feature.object_id}' found in 'features'!"
                raise ConfigError(msg)

            if feature.object_id:
                object_ids.append(feature.object_id)

    def get_cover_circuits(self) -> List[str]:
        """Get all circuits that are defined in the cover config."""
        circuits: List[str] = []

        for cover in self.covers:
            cover_up: str = cover.cover_up
            cover_down: str = cover.cover_down

            if cover_up:
                circuits.append(cover_up)

            if cover_down:
                circuits.append(cover_down)

        return circuits

    def _validate_covers_circuits(self) -> None:
        circuits: List[str] = self.get_cover_circuits()

        for circuit in circuits:
            if circuits.count(circuit) > 1:
                exception_message: str = (
                    f"{LogPrefix.COVER} Duplicate circuits found in 'covers'. "
                    f"Driving both signals up and down at the same time can damage the motor!"
                )
                raise ConfigError(exception_message)

    def _validate_cover_ids(self) -> None:
        object_ids: List[str] = []

        for cover in self.covers:
            if cover.object_id in object_ids:
                msg = f"{LogPrefix.COVER} Duplicate ID '{cover.object_id}' found in 'covers'!"
                raise ConfigError(msg)

            object_ids.append(cover.object_id)
